### PR TITLE
Stm32 serial improvements

### PR DESCRIPTION
**Pull request by LuchianMihai**
## Summary
Small reworks of stm32_serial.c and stm32_uart.h

## Impact
Minimal, should not alter current behavior of stm32_serial.c

## Testing
Locally, currently only on [nucleo-f429zi](https://www.st.com/en/evaluation-tools/nucleo-f429zi.html)

### CODE REVIEW

**Comment by davids5** [comment]
Please Build test this against a config for all the config cases that cover DMA, TX , RX, ALL ON, ALL OFF, 1 ON

**Comment by LuchianMihai** [comment]
Hi @davids5, Indeed, I have not touched DMA related stuff, TX was not compiling previously and I thought it was my configuration the issues, as for RX, it did build, but due to DMA half transfer and transfer complete interrupts, only after 16 chars sent they will get processed by the OS, so I have also avoided this case.

If you have any suggestions about how to deal with the DMA on console, I am happy to hear them.

**Comment by acassis** [comment]
> **L.E.:** I do not seem to be able to link this pull request to an issue, and how does the pipeline works? Is it triggered after every commit, or is manually triggered by reviewers, maintainers, contributors?

After two or three PRs approved manually by a committer with write permission, github will do it automatically for yours new PRs

File: arch/arm/src/stm32/hardware/stm32_dma_v2.h
Comment (davids5) [inline arch/arm/src/stm32/hardware/stm32_dma_v2.h:436]: This is a breaking change that adds no real value. - Consider dropping it.

File: arch/arm/src/stm32/stm32_uart.h
Comment (davids5) [inline arch/arm/src/stm32/stm32_uart.h:553]: Is CR3 on all the serial IP types of the SoCs? I.E. Not on an F1 but on F3 - etc

**Comment by davids5** [comment]
@LuchianMihai - 

Would you explain what you feel the benefits of the Serial change are?  Why passing a struct? Etc./

Also I see changes in when registers are read that will effect Reset on Read registers. That need to be tested.

**Review by davids5 — CHANGES_REQUESTED** [review]

**Comment by LuchianMihai** [comment]
> @LuchianMihai -
> 
> Would you explain what you feel the benefits of the Serial change are? Why passing a struct? Etc./
> 
Well, previously there was an uint16_t that handled the interaction with both CR1 and CR2, it worked by avoiding enabling overlapping interrupts. I think that there are better solution. Why a struct, because there are three control registers that can enable interrupts, so I thought of an solution to include all of them, and a struct made sense as it introduced least amount of changes, still a single parameter (for the functions interacting with usart interrupts) instead of two (cr1 & cr3, which are used), or three, if we take into consideration all control registers.
> Also I see changes in when registers are read that will effect Reset on Read registers. That need to be tested.
Indeed, my main issue was that we are storing in an struct values of an register, which is changed by the hardware, so most of the time those values (struct member and register) will be out of sink, and we anyway update the struct member value before using it, and in the case we are using the value without reading the register prior, I thought it is an good idea to avoid it altogether.
I've explicitly read the register everywhere I needed it's value, although I kept the same function logic, so there is no read outside the functions that were already reading those register, but indeed tests are needed. Please advise how to test my changes.

File: arch/arm/src/stm32/hardware/stm32_dma_v2.h
Comment (LuchianMihai) [reply arch/arm/src/stm32/hardware/stm32_dma_v2.h:436]: Indeed, I've had an deeper look, DMAMAP_USART3_TX should have been set/chosen in board.h, my board configuration did not have that. Will revert this change.

File: arch/arm/src/stm32/stm32_uart.h
Comment (LuchianMihai) [reply arch/arm/src/stm32/stm32_uart.h:553]: I've checked following SoCs: F0x0, F0x1, F0x4, F0x8, F100, F1x1, F1x2, F1x3, F1x5, F1x7, F301, F334, F37x, F405, F407, F401, F446. They all have all 3 control registers. G4xx have compatibility with F3. Skipped a few, If anything comes to mind, note it and I can check.

**Comment by acassis** [comment]
Hi @LuchianMihai why did you close your PR? You could just use git rebase -i to fix things

**Comment by LuchianMihai** [comment]
Hi @acassis, kinda bit more than I can chew for an first contribution.
I did not delete the branch locally and I am continuing the work on this changes.
But stm32 being an widely used, the testing suite is also pretty large, so my changes should me verified in detail. For now I cannot even start the pipeline to check my commits.
Currently I am working on an driver. Here no test will fail on new code :D.
I'll continue contributing until I can run my commit through the pipeline.
After that I'll either reopen this PR, or open a new one.
I've closed this PR just because I do not like pending/dangling PRs, a personal preference I think?
Maybe we need to do the same with my open Issue, it can be closed.

**Pull request closed without merging**
